mmcv/device/ipu/model_converter.py

```diff
@@ -0,0 +1,853 @@
+            static graph of model training or inference.
+        args (argument list): Arguments passed to the `__init__`
+            method of PoplarExecutor.
+        kwargs (keyword arguments): Keyword arguments passed to the `__init__`
+            method of PoplarExecutor.
+    """
+    def __init__(
+            self,
+            model,
+            logger=None,
+            training=True,
+            modules_to_record=None,
+            *args,
+            **kwargs
+            ):
+        # self.model == self._user_model: input pytorch model
+        # self._model: wrapped model which is used to compile
+        # and update weights, these two models use same weights
+        # wrapped model only accept and output tuple, so HierarchicalData
+        # will convert dictionary to tuple and convert them back
+        self.inputs_tree_manager = HierarchicalData(logger=logger)
+        self.outputs_tree_manager = HierarchicalData(logger=logger)
+        self.logger = logger
+        # the features calculated by CPU
+        self.inter_ouputs_in_cpu = {}
+        # the features calculated by IPU
+        self.inter_ouputs_in_ipu = {}
+        if modules_to_record is None:
+            # It is possible that the IPU implementation of some operators
+            # is inconsistent with the expected (CPU), here you can use
+            # this method to confirm whether there is a problem
+            self.compare_with_cpu = False
+        else:
+            self.compare_with_cpu = True
+        # move model.fp16_enabled to self.fp16_enabled,
+        # modify the position where the input is automatically casted to half
+        if getattr(model, 'fp16_enabled', False):
+            model.fp16_enabled = False
+            self.fp16_enabled = True
+        # make torch.jit.trace convert self._model
+        model = WrappedNet(model, self.inputs_tree_manager,
+                           self.outputs_tree_manager,
+                           self.inter_ouputs_in_cpu,
+                           modules_to_record=modules_to_record)
+        super().__init__(model, training=training, *args, **kwargs)
+        # overwrite self._args_parser in train_step or val_step
+        self._args_parser = None
+        if training:
+            assert self.training
+        else:
+            assert not self.training
+
+    @property
+    def training(self):
+        # If trying to get the attribute(training) of self,
+        # since the class has no training attribute,
+        # it will automatically look for the training attribute of self.model.
+        # However, the real attribute we want to check is self._training,
+        # self.model.training  and self._training are often inconsistent.
+        # It is not clear whether it is a Poptorch bug or a special design,
+        # temporarily use this function to fix the problem
+        return self._training  # comes from self.model._training
+
+    @auto_fp16(supported_types=(PoplarExecutor,))
```

Comment:
Should it be `nn.Module, PoplarExecutor`?

Reply:
it decorate PoplarExecutor's function, so just PoplarExecutor is okay